src/components/htmlComponent.ts

```diff
@@ -0,0 +1,247 @@
+
+import { Point, SpaceRequest, Bounds } from "../core/interfaces";
+import * as RenderController from "../core/renderController";
+import * as Utils from "../utils";
+
+export class HTMLComponent extends AbstractComponent<HTMLElement> {
+  private _cssClasses = new Utils.Set<string>();
+  private _element: HTMLDivElement;
+
+  public anchor(selection: HTMLElement) {
+    if (this._destroyed) {
+      throw new Error("Can't reuse destroy()-ed Components!");
+    }
+
+    if (this._element != null) {
+      // reattach existing element
+      selection.appendChild(this._element);
+    } else {
+      this._element = document.createElement("div");
+      selection.appendChild(this._element);
+      this._setup();
+    }
+
+
+    this._isAnchored = true;
+    this._onAnchorCallbacks.callCallbacks(this);
+    return this;
+  }
+
+  public anchorHTML(selection: HTMLElement) {
+    return this.anchor(selection);
+  }
+
+  public computeLayout(origin?: Point, availableWidth?: number, availableHeight?: number) {
+    if (origin == null || availableWidth == null || availableHeight == null) {
+      if (this._element == null) {
+        throw new Error("anchor() must be called before computeLayout()");
+      } else if (this.parent() == null) {
+        // if the parent is null we are the root node. In this case, we determine
+        // then sizing constraints for the remainder of the chart.
+
+        origin = { x: 0, y: 0 };
+        // this is the top-level element. make it 100% height and width
+        // so we can measure the amount of space available for the chart
+        // with respect to the total space allocated for charting as specified
+        // by the user.
+        this._element.style.width = "100%";
+        this._element.style.height = "100%";
+
+        availableWidth = Utils.DOM.elementWidth(this._element);
+        availableHeight = Utils.DOM.elementHeight(this._element);
+      } else {
+        throw new Error("null arguments cannot be passed to computeLayout() on a non-root node");
+      }
+    }
+
+    let size = this._sizeFromOffer(availableWidth, availableHeight);
+    this._width = size.width;
+    this._height = size.height;
+
+    let xAlignProportion = HTMLComponent._xAlignToProportion[this._xAlignment];
+    let yAlignProportion = HTMLComponent._yAlignToProportion[this._yAlignment];
+    this._origin = {
+      x: origin.x + (availableWidth - this.width()) * xAlignProportion,
+      y: origin.y + (availableHeight - this.height()) * yAlignProportion,
+    };
+
+    // set the size and position of the root element given the
+    // calculated space constraints
+    d3.select(this._element).style({
+      height: `${this._height}px`,
+      left: `${this._origin.x}px`,
+      top: `${this._origin.y}px`,
+      width: `${this._width}px`,
+    });
+
+    if (this._resizeHandler != null) {
+      this._resizeHandler(size);
+    }
+
+    return this;
+  }
+
+  public renderImmediately() {
+    return this;
```

Comment:
Sure, they can implement this if they want. Because this is an abstract method, the basic HTMLComponent has to do _something_ and it doesn't need to actually render itself.